### PR TITLE
PLANNER-2337 NPE in NearbyDistanceMatrix

### PR DESCRIPTION
**Pull request by rsynek**
### JIRA
https://issues.redhat.com/browse/PLANNER-2337

### Checklist
- [ ] Documentation updated if applicable.
- [ ] Upgrade recipe provided if applicable.

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>full downstream build</b> please add comment: <b>Jenkins run fdb</b>
* for a <b>compile downstream build</b> please add comment: <b>Jenkins run cdb</b>
* for a <b>full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>
* for an <b>upstream build</b> please add comment: <b>Jenkins run upstream</b>
* for a <b>Quarkus LTS check</b> please add comment: <b>Jenkins run LTS</b>
* for a <b>Native check</b> please add comment: <b>Jenkins run native</b>
</details>

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/nearby/NearbyDistanceMatrix.java [hunk 1]
```diff
@@ -20,20 +20,29 @@
...
 import java.util.Map;
+import java.util.function.Function;
+import java.util.function.ToIntFunction;
 
 public final class NearbyDistanceMatrix {
 
     private final NearbyDistanceMeter nearbyDistanceMeter;
     private final Map<Object, Object[]> originToDestinationsMap;
+    private final Function<Object, Iterator<Object>> destinationIteratorProvider;
+    private final ToIntFunction<Object> destinationSizeFunction;
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/nearby/NearbyDistanceMatrix.java:31]: That's a lot of `Object` there. Any chance to make the class generic?
Even if we can't use the generics on the outside, I think it still helps on the inside, makes the class better self-documenting I think.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/nearby/NearbyDistanceMatrix.java [hunk 2]
```diff
@@ -69,6 +78,14 @@ public void addAllDestinations(Object origin, Iterator<Object> destinationIterat
 
     public Object getDestination(Object origin, int nearbyIndex) {
         Object[] destinations = originToDestinationsMap.get(origin);
+        if (destinations == null) {
+            /*
+             * The item may be missing in the distance matrix due to an underlying filtering selector.
+             * In such a case, the distance matrix needs to be updated.
+             */
+            addAllDestinations(origin);
+            destinations = originToDestinationsMap.get(origin);
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/nearby/NearbyDistanceMatrix.java:87]: If it were me, I'd just recurse at this point - calling the same function on the new data, making sure it works.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/entity/nearby/NearEntityNearbyEntitySelector.java [hunk 3]
```diff
@@ -80,12 +80,22 @@ public void phaseStarted(AbstractPhaseScope<Solution_> phaseScope) {
...
                     + ") has an entitySize (" + childSize
                     + ") which is higher than Integer.MAX_VALUE.");
         }
+
+        nearbyDistanceMatrix = new NearbyDistanceMatrix(nearbyDistanceMeter, (int) originSize,
+                origin -> childEntitySelector.endingIterator(), origin -> computeDestinationSize(childSize));
+        for (Iterator<Object> originIt = replayingOriginEntitySelector.endingIterator(); originIt.hasNext();) {
+            final Object origin = originIt.next();
+            nearbyDistanceMatrix.addAllDestinations(origin);
+        }
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/entity/nearby/NearEntityNearbyEntitySelector.java:95]: FYI there is a `forEachRemaining(...)` method on an `Iterator`.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/value/nearby/NearEntityNearbyValueSelector.java [hunk 4]
```diff
@@ -78,28 +78,33 @@ public void phaseStarted(AbstractPhaseScope<Solution_> phaseScope) {
...
+    private int computeDestinationSize(Object origin) {
+        long childSize = childValueSelector.getSize(origin);
+        if (childSize > (long) Integer.MAX_VALUE) {
+            throw new IllegalStateException("The childEntitySelector (" + childValueSelector
+                    + ") has an entitySize (" + childSize
+                    + ") which is higher than Integer.MAX_VALUE.");
+        }
+        int destinationSize = (int) childSize;
+        if (randomSelection) {
+            // Reduce RAM memory usage by reducing destinationSize if nearbyRandom will never select a higher value
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/value/nearby/NearEntityNearbyValueSelector.java:98]:
```suggestion
            // Reduce heap usage by reducing destinationSize if nearbyRandom will never select a higher value.
```

**Review by triceo — APPROVED** [review]
I like the coverage! Left a few suggestions for improvement.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/nearby/NearbyDistanceMatrix.java (hunk 2 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/nearby/NearbyDistanceMatrix.java:87]: +0 if, for any reason, the matrix is not updated, the current approach ends up with NPE, the approach you're suggesting ends up with StackOverflowError.

Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/nearby/NearbyDistanceMatrix.java:87]: Fair enough.

**Pull request merged**